pkg/actors/actors.go

```diff
@@ -1036,12 +1153,243 @@ func (a *actorsRuntime) executeTimer(actorType, actorID, name, dueTime, period,
+func (a *actorsRuntime) saveActorTypeMetadata(actorType string, actorMetadata *ActorMetadata) error {
+	if !a.actorTypeMetadataEnabled {
+		return nil
+	}
+
+	metadataKey := constructCompositeKey("actors", actorType, "metadata")
+	return a.store.Set(&state.SetRequest{
+		Key:   metadataKey,
+		Value: actorMetadata,
+		ETag:  actorMetadata.Etag,
+	})
+}
+
+func (a *actorsRuntime) getActorTypeMetadata(actorType string, migrate bool) (*ActorMetadata, error) {
+	if a.store == nil {
+		return nil, errors.New("actors: state store does not exist or incorrectly configured")
+	}
+
+	if !a.actorTypeMetadataEnabled {
+		return &ActorMetadata{
+			ID: uuid.NewString(),
+			RemindersMetadata: ActorRemindersMetadata{
+				partitionsEtag: nil,
+				PartitionCount: 0,
+			},
+			Etag: nil,
+		}, nil
+	}
+
+	metadataKey := constructCompositeKey("actors", actorType, "metadata")
+	resp, err := a.store.Get(&state.GetRequest{
+		Key: metadataKey,
+	})
+	if err != nil || len(resp.Data) == 0 {
+		// Metadata field does not exist or failed to read.
+		// We fallback to the default "zero" partition behavior.
+		actorMetadata := ActorMetadata{
+			ID: uuid.NewString(),
+			RemindersMetadata: ActorRemindersMetadata{
+				partitionsEtag: nil,
+				PartitionCount: 0,
+			},
+			Etag: nil,
+		}
+
+		// Save metadata field to make sure the error was due to record not found.
+		// If the previous error was due to database, this write will fail due to:
+		//   1. database is still not responding, or
+		//   2. etag does not match since the item already exists.
+		// This write operation is also needed because we want to avoid a race condition
+		// where another sidecar is trying to do the same.
+		etag := ""
+		if resp != nil && resp.ETag != nil {
+			etag = *resp.ETag
+		}
+
+		actorMetadata.Etag = &etag
+		err = a.saveActorTypeMetadata(actorType, &actorMetadata)
+		if err != nil {
+			return nil, err
+		}
+
+		// Needs to read to get the etag
+		resp, err = a.store.Get(&state.GetRequest{
+			Key: metadataKey,
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var actorMetadata ActorMetadata
+	err = json.Unmarshal(resp.Data, &actorMetadata)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse metadata for actor type %s (%s): %w", actorType, string(resp.Data), err)
+	}
+	actorMetadata.Etag = resp.ETag
+	if !migrate {
+		return &actorMetadata, nil
+	}
+
+	return a.migrateRemindersForActorType(actorType, &actorMetadata)
+}
+
+func (a *actorsRuntime) migrateRemindersForActorType(actorType string, actorMetadata *ActorMetadata) (*ActorMetadata, error) {
+	if !a.actorTypeMetadataEnabled {
+		return actorMetadata, nil
+	}
+
+	if actorMetadata.RemindersMetadata.PartitionCount == a.config.RemindersStoragePartitions {
+		return actorMetadata, nil
+	}
+
+	if actorMetadata.RemindersMetadata.PartitionCount > a.config.RemindersStoragePartitions {
+		log.Warnf("cannot decrease number of partitions for reminders of actor type %s", actorType)
```

Comment:
Any reason why we don't support a decrementing path? I know it's not a hard one-way door as you can always delete and recreate your reminders but this feels like something we can support without much extra effort and it makes it easier for a user to walk back their changes.

Reply:
Actually, this is an important design decision. If we allow both increment and decrement scenarios, we will be in a ambiguous migration where instances running with previous version will be migrating with decrement path while instances running with the new version will be migrating with the increment path. It would be an infinite loop. By forcing only the increment path, it removes ambiguity and makes sure that only one migration happens per cluster.

Reply:
In the case of the previous version, wouldn't it not have the migration code?  Though it does bring up a good point. What is the rollback scenario here? If you upgrade to 1.3, have a reminder trigger the migration, and then rollback all of your reminders are no longer reachable, right? Or at least a subset of them?

What I was trying to say though was a user can reset their actor type to have a lower specified type. It seems like in that case, the ETag barrier should stop us from getting into a loop, right?

Again, the main worry I have here is just how to do we back away from this change if something goes wrong (even unrelated to this, but in 1.3 as a whole). Or, if someone makes a mistake and provisions many more partitions than required and gets into a degraded experience, is there a way out?

Reply:
First, there is no rollback to a previous version if user enabled partitioning. This is the main reason this is an opt-in feature. If user wants to rollback from 1.3 to 1.2 it will work as usual if this feature was not used. There is no way out of this because we cannot keep both formats in parallel.

What trigger the migration code is the fact that the sidecar detects "desired number of partitions is not equal to current number of partitions" - so I block the decrement scenario to avoid different versions of the sidecar configuration to migrate back and forth.

Reply:
Gotcha. As long as we're making the conscious decision and it's opt-in I'm fine with that.

I do think we could design around this and get to a place where up/down/rollback is supported but it'd add complexity to it which I don't think we have a reason given that it's an opt-in preview feature.

Reply:
It would add a new level of complexity and testing. If we need to cross that bridge, I would rather implement a generic metadata storage API in dapr.

Reply:
Agreed with @artursouza.